### PR TITLE
Manual GC support pointer implmentation

### PR DESCRIPTION
**Pull request by arccth**

### CODE REVIEW

File: course4/homework-unique-pointer/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
+
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/CMakeLists.txt:4]: Good; much more portable that the manual param tweaking I tend to use. If only there were a way to enable warnings the same way...

File: course4/homework-unique-pointer/collectable.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+public:
+    CollectablePointer(const CollectablePointer& other)
+    {
+        payload = new T(*other.payload);
+        std::cout << "allocated as copy: " << (void*)payload << std::endl;
+    }
+
+    CollectablePointer(CollectablePointer&& other)
+    {
+        payload = std::move(other.payload);
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/collectable.h]: what is the difference between simply copying the pointer, and std:moving it?

File: course4/homework-unique-pointer/collectable.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    static CollectablePointer<T> createInstance(TArg arg)
+    {
+        CollectablePointer<T> result;
+        result.payload = new T(arg);
+        std::cout << "allocated(arg): "<<arg<<" " << (void*)result.payload << std::endl;
+        return result;
+    }
+
+    template<typename TArg1, typename TArg2>
+    static CollectablePointer<T> createInstance(TArg1 arg1, TArg2 arg2)
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/collectable.h]: What you really need here, are variadic templates.

File: course4/homework-unique-pointer/test.cc [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+{
+    auto c1 = course::CollectablePointer<T>::createInstance(arg1);
+    std::cout << "Done 1" << std::endl;
+    auto c2 = course::CollectablePointer<T>::createInstance(arg2);
+    std::cout << "Done 2" << std::endl;
+    auto c3 = c1;
+    std::cout << "Done 3" << std::endl;
+    auto c4 = std::move(c2);
+    std::cout << "Done 4" << std::endl;
+    assert(course::CollectablePointer<T>::itemsToCollectCount() == 0);
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/test.cc:14]: shouldn't this be 3?

File: course4/homework-unique-pointer/collectable.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    ~CollectablePointer()
+    {
+        if(payload != nullptr)
+        {
+            CollectablePointer<T>::poolOfDead.push_back(payload);
+            payload = nullptr;
+        }
+    }
+
+    T& get()
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/collectable.h:36]: Well design smart pointer should have intereface as similar to a raw pointer as possible.
Therefore it should have at least overloaded operator `*` and `->`.

File: course4/homework-unique-pointer/test.cc [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    course::CollectablePointer<T>::collect();
+    assert(course::CollectablePointer<T>::itemsToCollectCount() == 0);
+}
+
+int main()
+{
+    auto c1 = course::CollectablePointer<std::string>::createInstance("TEST");
+    assert(c1.get() == "TEST");
+
+    std::cout<<"Start refcount tests" << std::endl;
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/test.cc]: There is no refcounting in there?

**Review by maciekgajewski — COMMENTED** [review]
Interesting.

It would good to add some documentation; for instance: what would be the benefits from using this class.

It's difficult to call this a "smart pointer", as it doesn't have the pointer interface.

File: course4/homework-unique-pointer/collectable.h (hunk 5 above)
Comment (arccth) [reply course4/homework-unique-pointer/collectable.h:36]: Done

File: course4/homework-unique-pointer/collectable.h (hunk 2 above)
Comment (arccth) [reply course4/homework-unique-pointer/collectable.h]: Changed to simple copy

File: course4/homework-unique-pointer/collectable.h (hunk 3 above)
Comment (arccth) [reply course4/homework-unique-pointer/collectable.h]: Done

File: course4/homework-unique-pointer/test.cc (hunk 4 above)
Comment (arccth) [reply course4/homework-unique-pointer/test.cc:14]: No, at this moment destructors are not called yet, so no garbage

**Comment by arccth** [comment]
Idea of this pointer is to postpone delete operations from performance critical section, and then manually collect all garbage

**Comment by arccth** [comment]
Added refcount pointer

File: course4/homework-unique-pointer/refcounted.h [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+#include<iostream>
+
+namespace course{
+
+template <typename T>
+class RefcountingPointer
+{
+private:
+    T* payload;
+    uint* count;
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/refcounted.h]: I'm not sure if 'uint' is standard type

File: course4/homework-unique-pointer/refcounted.h [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    void decrement()
+    {
+        if(count == nullptr)
+        {
+            return;
+        }
+
+        if(--(*count) == 0)
+        {
+            std::cout << "Deleting: " << (void*)payload << std::endl;
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/refcounted.h]: These are good for debugging during development, but you should remove them when you are done.
You should be able to wrtie automatic unit-test.

File: course4/homework-unique-pointer/test.cc [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+{
+    innerCall<T, Arg>(arg1, arg2);
+
+    assert(course::CollectablePointer<T>::itemsToCollectCount() == 3);
+    course::CollectablePointer<T>::collect();
+    assert(course::CollectablePointer<T>::itemsToCollectCount() == 0);
+}
+
+template<typename T, typename Arg>
+void refcauntTest(Arg arg)
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/test.cc]: refcount

File: course4/homework-unique-pointer/test.cc [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+     assert(c1.getCount() == 3);
+     auto c4 = std::move(c2);
+     assert(*c4 == *c3);
+     assert(c1.getCount() == 3);
+     c2 = c1;
+     assert(c1.getCount() == 4);
+     c3 = c1;
+     assert(c1.getCount() == 4);
+     c2 = std::move(c4);
+     assert(c1.getCount() == 3);
```
Comment (maciekgajewski) [inline course4/homework-unique-pointer/test.cc:43]: Good test, but incomplete.

Create an class that modifies some global variable (or variables) in constructor and destructor. This way you could verify if object is destroyed when you expect it to be.

**Review by maciekgajewski — COMMENTED** [review]
- Remvoe debug printing
- Run the code trough valgrind, I expect you would see some memory errors.

File: course4/homework-unique-pointer/test.cc (hunk 10 above)
Comment (arccth) [reply course4/homework-unique-pointer/test.cc:43]: Done

File: course4/homework-unique-pointer/refcounted.h (hunk 8 above)
Comment (arccth) [reply course4/homework-unique-pointer/refcounted.h]: Done

**Comment by arccth** [comment]
Finally valgrind output 0 errors from 0 contexts (suppressed: 0 from 0)

**Pull request closed without merging**